src/aspire/denoising/denoiser_cov2d.py

```diff
@@ -18,72 +20,115 @@ class DenoiserCov2D(Denoiser):
+                                                cov2d.ctf_idx[img_start:img_start+batch_size[ibatch]],
+                                                mean_coeff=mean_coeffs_est, covar_coeff=covar_coeffs_est,
+                                                noise_var=self.var_noise)
 
-        covar_coeffs_est = cov2d.get_covar(self.coeffs_noise, self.ctf_fb, self.ctf_idx, mean_coeffs_est,
-                                               noise_var=self.var_noise, covar_est_opt=covar_opt)
+            # Convert Fourier-Bessel coefficients back into 2D images
+            logger.info(f'Converting Cov2D coefficients back to 2D images')
+            imgs_estim = self.basis.evaluate(coeffs_estim)
+            imgs_out = Image(imgs_estim)
+            mrcs_file = os.path.join(self.src.proj_folder, self.starfile_path, self.mrcs_out[img_start])
+            logger.info(f'Saving denoised images to {mrcs_file}')
+            imgs_out.save(mrcs_file, overwrite=overwrite)
+            img_start += batch_size[ibatch]
+
+    def _write_star_file(self):
+        """
+        Save a new star file with field for denoised images
+        """
 
-        self.coeffs_estim = cov2d.get_cwf_coeffs(self.coeffs_noise, self.ctf_fb, self.ctf_idx,
-                                                     mean_coeff=mean_coeffs_est,
-                                                     covar_coeff=covar_coeffs_est, noise_var=self.var_noise)
+        self.metadata['_rlnDenoisedImageName'] = pd.Series('', index=self.metadata.index)
+        self.metadata = self.metadata.drop(['__mrc_filename', '__mrc_index', '__mrc_filepath',
+                                            '__filter', '__filter_indices'], axis=1)
+        df = self.metadata
+        for idx in range(len(df.index)):
+            img_prefix = df['_rlnImageName'][idx].split('@')[0]
+            mrcs_filein = df['_rlnImageName'][idx].split('/')[-1]
+            mrcs_fileout = mrcs_filein.split('.')[0] + f'_denoised.mrcs'
+            mrcs_filepath = img_prefix + '@' + os.path.join(self.starfile_path, mrcs_fileout)
+            df.loc[idx, '_rlnDenoisedImageName'] = mrcs_filepath
 
-        # Convert Fourier-Bessel coefficients back into 2D images
-        self.imgs_estim = self.basis.evaluate(self.coeffs_estim)
+        with open(os.path.join(self.src.proj_folder, self.star_file), 'w') as f:
+            starfile = StarFile(blocks=[StarFileBlock(loops=[df])])
+            starfile.save(f)
 
-    def to_src(self):
+    def _get_mrcs_out(self):
         """
-        Return an ImageSource object with denoised images
+        Generate a name list of output mrcs files for denoised images
         """
-        src = ArrayImageSource(Image(self.imgs_estim))
-        return src
+        df = self.metadata
+        mrcs_fileout = []
+        for idx in range(len(df.index)):
+            mrcs_fileout.append(df['_rlnDenoisedImageName'][idx].split('/')[-1])
```

Comment:
Let's not use new fields unless it's absolutely necessary. This won't be interpretable by RELION. Better to store it in `_rlnImageName`.

Reply:
It should be fine to use the same name if we just replace the original noise image.

Reply:
Change to use the same name `_rlnImageName` in the new star file.